src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/CastCache.cs

```diff
@@ -15,47 +18,52 @@ internal enum CastResult
         MaybeCast = 2
     }
 
-#if NATIVEAOT
-    [EagerStaticClassConstruction]
-#endif
-    internal static unsafe class CastCache
+    internal unsafe struct CastCache
     {
-
-#if CORECLR
-        // In coreclr the table is written to only on the native side. T
-        // This is all we need to implement TryGet.
-        private static int[]? s_table;
-#else
-
-  #if DEBUG
-        private const int INITIAL_CACHE_SIZE = 8;    // MUST BE A POWER OF TWO
-        private const int MAXIMUM_CACHE_SIZE = 512;  // make this lower than release to make it easier to reach this in tests.
-  #else
-        private const int INITIAL_CACHE_SIZE = 128;  // MUST BE A POWER OF TWO
-        private const int MAXIMUM_CACHE_SIZE = 4096; // 4096 * sizeof(CastCacheEntry) is 98304 bytes on 64bit. We will rarely need this much though.
-  #endif // DEBUG
-
         private const int VERSION_NUM_SIZE = 29;
         private const uint VERSION_NUM_MASK = (1 << VERSION_NUM_SIZE) - 1;
+        private const int BUCKET_SIZE = 8;
+
+        // nothing is ever stored into this, so we can use a static instance.
+        private static int[]? s_sentinelTable;
 
-        // A trivial 2-elements table used for "flushing" the cache. Nothing is ever stored in this table.
-        // It is required that we are able to allocate this.
-        private static int[] s_sentinelTable = CreateCastCache(2, throwOnFail: true)!;
+        // The actual storage.
+        private int[] _table;
 
         // when flushing, remember the last size.
-        private static int s_lastFlushSize = INITIAL_CACHE_SIZE;
+        private int _lastFlushSize;
 
-        // The actual storage.
-        // Initialize to the sentinel in DEBUG as if just flushed, to ensure the sentinel can be handled in Set.
-        private static int[] s_table =
-  #if !DEBUG
-            CreateCastCache(INITIAL_CACHE_SIZE) ??
-  #endif
-            s_sentinelTable;
+        private int _initialCacheSize;
+        private int _maxCacheSize;
 
-#endif // CORECLR
+        // wraps existing table
```

Comment:
How does this wrapping work? It won't intialize the other fields like `_lastFlushSize`, etc.

Reply:
I looked at jitted code and if I am not mistaken it looks like the struct gets promoted and unused fields disappear. When we operate with a cache that wraps an existing table, we only do `TryGet`, and `TryGet` only needs the `_table`.

I looked at the code in the debugger, which is not so trivial because of "no debugging this" attributes. There is a strange call there, most likely related to accessing a static field, but otherwise the codegen looked like the struct gets completely erased. 

It would be nice if someone could confirm that this is the case. There might be a "proper" way to just dump the jitted methods and see what happens there.

Reply:
I am more worried about that this constructor creates `CastCache` instance that is partially initialized. It makes it hard to see that none of the places that use this instance won't use the uninitialized fields (`_maxCacheSize`, etc.).

Reply:
The code paths that add/resize/flush are not supposed to be used in a cache that wraps existing table since logically they do not own the table. An eventual crash due to uninitialized fields would be expected. 

I guess we could have more explicit asserts - if you are off the TryGet path the min/max sizes should be nonzero.

Reply:
I would be better to structure the code such that these partially initialized `CastCache` instances do not need to be created.

Reply:
> I would be better to structure the code such that these partially initialized CastCache instances do not need to be created.

That can be done if we pass the table directly to `TryGet`. I think that will work better.We would not need to add asserts and worry about the struct being optimized away.

Reply:
I’ve removed the wrapping scenario. Just pass the existing table, if available, to ‘TryGet’.